### DISCUSSION TITLE
set linearization_level doesn't work any more

### DISCUSSION THREAD

**Discussion by C8504** [google/or-tools#2804] [CP-SAT questions] [answered]
Hi Experts,

I just tried the latest v9.1, found linearization_level parameter in SAT doesn't work in Python any more,
could you pelase take a look and confirm?

thanks!

**Comment by lperron** [comment]
What are the other parameters ?
What is the symptom?

**Reply by C8504** [reply]
```python
from ortools.linear_solver import pywraplp
solver = pywraplp.Solver("SOLVER", pywraplp.Solver.SAT_INTEGER_PROGRAMMING)
solver.SetSolverSpecificParametersAsString("linearization_level:0")
solver.SetNumThreads(1)

x = solver.BoolVar('x')
y = solver.IntVar(0, 2, 'y')

print('Number of variables =', solver.NumVariables())

ct = solver.Constraint(0, 2, 'ct')
ct.SetCoefficient(x, 1)
ct.SetCoefficient(y, 1)

print('Number of constraints =', solver.NumConstraints())

objective = solver.Objective()
objective.SetCoefficient(x, 3)
objective.SetCoefficient(y, 1)
objective.SetMaximization()

solver.EnableOutput()
solver.Solve()
```

If you run above sample, set linearization_level to 0, while the log show 

`Starting CP-SAT solver.
Parameters: log_search_progress: true linearization_level: 2 num_search_workers: 1`

**Comment by lperron** [comment, marked as answer]
Indeed. Thanks for the report

**Reply by lperron** [reply]
Converted to issue.

**Reply by lperron** [reply]
and fixed on master.  Thanks again.

**Reply by C8504** [reply]
You are welcome!

**Reply by C8504** [reply]
Hello experts, 

In the latest v 9.3, linearization_level parameter doesn't work again, could you please take a look? thanks a lot!

John

**Reply by lperron** [reply]
Not sure:

```
solve --input ~/Work/support/hs_lp.bin  --solver sat --params num_search_workers:1,linearization_level:0
```

->

```
 Read input proto as an MPModelProto.
File        : '/Users/lperron/Work/support/hs_lp.bin'
I0311 09:14:28.525358     1 solve.cc:179] Set number of threads to 1.
Solver      : SAT_INTEGER_PROGRAMMING
Dimension   : 104155 x 39247

Running basic LP presolve, initial problem dimensions: 104155 rows, 39247 columns, 157736 entries with magnitude in [3.264389e-02, 3.186730e+03]
glop::FixedVariablePreprocessor                        104155 rows, 39247 columns, 157736 entries with magnitude in [3.264389e-02, 3.186730e+03]
glop::SingletonPreprocessor                            38834 rows, 39247 columns, 92415 entries with magnitude in [3.264389e-02, 3.186730e+03]
glop::ForcingAndImpliedFreeConstraintPreprocessor      38834 rows, 39247 columns, 92415 entries with magnitude in [3.264389e-02, 3.186730e+03]
glop::FreeConstraintPreprocessor                       36506 rows, 39247 columns, 87169 entries with magnitude in [3.264389e-02, 3.186730e+03]
glop::UnconstrainedVariablePreprocessor                36506 rows, 39247 columns, 87169 entries with magnitude in [3.264389e-02, 3.186730e+03]

Scaling to pure integer problem.
Num integers: 0/39247 (implied: 0 in_inequalities: 0 max_scaling: 0) [MIP] 
Warning: 5354 bounds were truncated to 10000000.
Warning: 2588 continuous variable domain with fewer than 1000 values.
Maximum constraint coefficient relative error: 1.35866e-05
Maximum constraint worst-case activity error: 180.867 [Potentially IMPRECISE]
Maximum constraint scaling factor: 8.58993e+09

Starting CP-SAT solver v9999.0.0
Parameters: log_search_progress: true linearization_level: 0 num_search_workers: 1
```

**Reply by C8504** [reply]
Thanks lperron,

I set linearization_level is 0, and num_search_workers set to 2, then I got below log which contains defult_lp solver, is not compability with older versions.


Starting CP-SAT solver v9.3.10459
Parameters: log_search_progress: true linearization_level: 0 num_search_workers: 2

......


Starting Search at 0.01s with 2 workers.
1 full subsolvers: [default_lp]
Interleaved subsolvers: [rnd_var_lns_default, rnd_cst_lns_default, graph_var_lns_default, graph_cst_lns_default, rins_lns_default, rens_lns_default]
#1       0.01s best:6     next:[]         default_lp fixed_bools:0/0
#Done    0.01s default_lp

John

**Comment by lperron** [comment]
Indeed. We need to decide what to do.

Thanks
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by C8504** [reply]
Thanks

**Comment by lperron** [comment]
So, conclusion. We have frozen most of the workers' parameters. In
particular, default_lp is linearization_level=1.
But, we have added a parameter to specify which worker types to use. So in
your case, you will be able to specify you want to use the no_lp instead of
the default lp.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
